File: dotnet/RailsLib.Net/Net/Algorithms/NetworkVertex.cs
using GameLib.Net.Common;
using GameLib.Net.Game;
using GameLib.Net.Graph;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Text;

namespace GameLib.Net.Algorithms
{
    public class NetworkVertex : IComparable<NetworkVertex>, IEquatable<NetworkVertex>
    {
        protected static Logger<NetworkVertex> log = new Logger<NetworkVertex>();

        public enum VertexTypes
        {
            STATION,
            SIDE,
            HQ,
        }

        public enum StationTypes
        {
            MAJOR,
            MINOR
        }

        // vertex types and flag for virtualVertex (thus not related to a rails object)
        private VertexTypes type;
        private bool virtualVertex;

        // vertex properties (for virtualVertex vertexes)
        private string virtualId;

        // general vertex properties
        private StationTypes stationType;
        private int value = 0;
        private bool sink = false;
        private string stopName = null;

        // references to rails objects, if not virtualVertex
        private MapHex hex;
        private TrackPoint trackPoint;
        // only for station objects
        private Stop stop;

        /** constructor for station on mapHex */
        public NetworkVertex(MapHex hex, Station station)
        {
            this.type = VertexTypes.STATION;
            this.hex = hex;
            this.trackPoint = station;
            this.stop = hex.GetRelatedStop(station);
            if (stop != null)
            {
                log.Info("Found stop " + stop);
            }
            else
            {
                log.Info("No stop found");
            }

            this.virtualVertex = false;
            this.virtualId = null;
        }

        /** constructor for side on mapHex */
        public NetworkVertex(MapHex hex, HexSide side)
        {
            this.type = VertexTypes.SIDE;
            this.hex = hex;
            this.trackPoint = side;
            this.stop = null;

            this.virtualVertex = false;
            this.virtualId = null;
        }

        /**  constructor for public company hq */
        public NetworkVertex(PublicCompany company) : this(VertexTypes.HQ, "HQ")
        {

        }

        private NetworkVertex(VertexTypes type, string name)
        {
            this.type = type;
            this.hex = null;
            this.trackPoint = null;
            this.stop = null;

            this.virtualVertex = true;
            this.virtualId = name;
        }

        /** factory method for virtualVertex vertex
         */
        public static NetworkVertex GetVirtualVertex(VertexTypes type, string name)
        {
            NetworkVertex vertex = new NetworkVertex(type, name);
            return vertex;
        }

        public void AddToRevenueCalculator(RevenueCalculator rc, int vertexId)
        {
            rc.SetVertex(vertexId, IsMajor, IsMinor, sink);
        }

        public string Identifier
        {
            get
            {
                if (virtualVertex)
                {
                    return virtualId;
                }
                else
                {
                    return hex.Id + "." + trackPoint.TrackPointNumber;
                }
            }
        }

        public bool IsVirtual
        {
            get
            {
                return virtualVertex;
            }
        }

        public bool IsStation
        {
            get
            {
                return type == VertexTypes.STATION;
            }
        }

        public bool IsSide
        {
            get
            {
                return type == VertexTypes.SIDE;
            }
        }

        public bool IsHQ
        {
            get
            {
                return type == VertexTypes.HQ;
            }
        }

        public VertexTypes VertexType
        {
            get
            {
                return type;
            }
        }

        public bool IsMajor
        {
            get
            {
                return (/*stationType != null &&*/ stationType == StationTypes.MAJOR);
            }
        }

        public bool IsMinor
        {
            get
            {
                return (/*stationType != null &&*/ stationType == StationTypes.MINOR);
            }
        }

        public StationTypes StationType
        {
            get
            {
                return stationType;
            }
            set
            {
                stationType = value;
            }
        }

        public NetworkVertex SetStationType(StationTypes stationType)
        {
            this.stationType = stationType;
            return this;
        }

        public int Value
        {
            get
            {
                return value;
            }
            set
            {
                this.value = value;
            }
        }

        public int GetValueByTrain(NetworkTrain train)
        {
            int valueByTrain;
            if (IsMajor)
            {
                valueByTrain = value * train.MultiplyMajors;
            }
            else if (IsMinor)
            {
                if (train.IgnoresMinors)
                {
                    valueByTrain = 0;
                }
                else
                {
                    valueByTrain = value * train.MultiplyMinors;
                }
            }
            else
            {
                valueByTrain = value;
            }
            return valueByTrain;
        }

        public NetworkVertex SetValue(int value)
        {
            this.value = value;
            return this;
        }

        public bool IsSink
        {
            get
            {
                return sink;
            }
            set
            {
                sink = value;
            }
        }

        //public NetworkVertex SetSink(bool sink)
        //{
        //    this.sink = sink;
        //    return this;
        //}

        public string StopName
        {
            get
            {
                return stopName;
            }
        }

        public NetworkVertex SetStopName(string locationName)
        {
            this.stopName = locationName;
            return this;
        }

        // getter for rails objects
        public MapHex Hex
        {
            get
            {
                return hex;
            }
        }

        public Station Station
        {
            get
            {
                if (type == VertexTypes.STATION)
                {
                    return (Station)trackPoint;
                }
                else
                {
                    return null;
                }
            }
        }

        public HexSide Side
        {
            get
            {
                if (type == VertexTypes.SIDE)
                {
                    return (HexSide)trackPoint;
                }
                else
                {
                    return null;
                }
            }
        }

        public Stop Stop
        {
            get
            {
                return stop;
            }
        }

        public bool IsOfType(VertexTypes vertexType, StationTypes stationType)
        {
            return (type == vertexType && (!IsStation || StationType == stationType));
        }

        /**
         * Initialize for rails vertexes
         * @return true = can stay inside the network, false = has to be removed
         */
        public bool InitRailsVertex(PublicCompany company)
        {
            // side vertices use the defaults, virtuals cannot use this function
            if (virtualVertex || type == VertexTypes.SIDE) return true;

            // Only station remains
            Station station = (Station)trackPoint;

            log.Info("Init of vertex " + this);

            // check if it has to be removed because it is run-to only
            // if company == null, then no vertex gets removed
            if (company != null && !stop.IsRunToAllowedFor(company))
            {
                log.Info("Vertex is removed");
                return false;
            }

            // check if it is a major or minor
            if (stop.ScoreType == StopType.Score.MAJOR)
            {
                SetStationType(StationTypes.MAJOR);
            }
            else if (stop.ScoreType == StopType.Score.MINOR)
            {
                SetStationType(StationTypes.MINOR);
            }

            // check if it is a sink
            if (company == null)
            { // if company == null, then all sinks are deactivated
                sink = false;
            }
            else
            {
                sink = !stop.IsRunThroughAllowedFor(company);
            }

            // define locationName
            stopName = null;
            if (station.GetStationType() == Station.StationType.OFFMAPCITY)
            {
                if (hex.StopName != null && !hex.StopName.Equals(""))
                {
                    stopName = hex.StopName;
                }
            }
            else
            {
                if (hex.StopName != null && !hex.StopName.Equals("")
                        && station.StopName != null && !station.StopName.Equals(""))
                {
                    stopName = hex.StopName + "." + station.StopName;
                }
            }

            // no removal
            return true;

        }

        public void SetRailsVertexValue(Phase phase)
        {
            // side vertices and  virtuals cannot use this function
            if (virtualVertex || type == VertexTypes.SIDE) return;

            // define value
            value = stop.GetValueForPhase(phase);
        }


        override public string ToString()
        {
            StringBuilder message = new StringBuilder();
            if (IsVirtual)
                message.Append(virtualId);
            else if (IsStation)
                message.Append(hex.Id + "." + ((Station)trackPoint).Number);
            else if (IsSide)
                message.Append(hex.Id + "." + hex.GetOrientationName((HexSide)trackPoint));
            else
                message.Append("HQ");
            if (IsSink)
                message.Append("/*");
            return message.ToString();
        }

        public int CompareTo(NetworkVertex otherVertex)
        {
            return this.Identifier.CompareTo(otherVertex.Identifier);
        }

        public bool Equals(NetworkVertex other)
        {
            if (other == null) return false;
            return Identifier == other.Identifier;
        }

        public override bool Equals(object obj)
        {
            if (obj == null) return false;
            if (!(obj is NetworkVertex)) return false;
            return Identifier == ((NetworkVertex)obj).Identifier;
        }

        public override int GetHashCode()
        {
            return Identifier.GetHashCode();
        }

        public class ValueOrder : IComparer<NetworkVertex>
        {
            public int Compare(NetworkVertex vA, NetworkVertex vB)
            {
                int result = -vA.Value.CompareTo(vB.Value); // compare by value, descending
                if (result == 0)
                    result = vA.CompareTo(vB); // otherwise use natural ordering
                return result;
            }
        }

        /**
         * 
         * @param graph network graph
         * @param company the company (with regard to values, sinks and removals)
         * @param phase the current phase (with regard to values)
         */
        public static void InitAllRailsVertices(NetworkGraph graph,
                PublicCompany company, Phase phase)
        {

            // store vertices for removal
            List<NetworkVertex> verticesToRemove = new List<NetworkVertex>();
            foreach (NetworkVertex v in graph.Graph.Vertices)
            {
                if (company != null)
                {
                    if (!v.InitRailsVertex(company))
                    {
                        verticesToRemove.Add(v);
                    }
                }
                if (phase != null)
                {
                    v.SetRailsVertexValue(phase);
                }
            }
            graph.Graph.RemoveVertices(verticesToRemove);
        }

        /**
         * Returns the maximum positive value (lower bound zero)
         */
        public static int MaxVertexValue(IEnumerable<NetworkVertex> vertices)
        {
            int maximum = 0;
            foreach (NetworkVertex vertex in vertices)
            {
                maximum = Math.Max(maximum, vertex.Value);
            }
            return maximum;
        }


        /**
         * Return the sum of vertex values
         */
        public static int Sum(IEnumerable<NetworkVertex> vertices)
        {
            int sum = 0;
            foreach (NetworkVertex vertex in vertices)
            {
                sum += vertex.Value;
            }
            return sum;
        }

        /**
         * Returns the number of specified vertex type in a vertex collection
         * If station then specify station type
         */
        public static int NumberOfVertexType(IEnumerable<NetworkVertex> vertices, VertexTypes vertexType, StationTypes stationType)
        {
            int number = 0;
            foreach (NetworkVertex vertex in vertices)
            {
                if (vertex.IsOfType(vertexType, stationType)) number++;
            }
            return number;
        }

        /**
         * creates a new virtualVertex vertex with identical properties and links
         */
        public static NetworkVertex DuplicateVertex(SimpleGraph<NetworkVertex, NetworkEdge> graph,
                NetworkVertex vertex, string newIdentifier, bool addOldVertexAsHidden)
        {
            // create new vertex
            NetworkVertex newVertex = NetworkVertex.GetVirtualVertex(vertex.type, newIdentifier);
            // copy values
            newVertex.stationType = vertex.stationType;
            newVertex.value = vertex.value;
            newVertex.sink = vertex.sink;
            newVertex.stopName = vertex.stopName;
            graph.AddVertex(newVertex);
            // copy edges
            //HashSet<NetworkEdge> edges = graph.EdgesOf(vertex);
            var edges = graph.EdgesOf(vertex);
            foreach (NetworkEdge edge in edges)
            {
                List<NetworkVertex> hiddenVertices;
                if (edge.Source == vertex)
                {
                    hiddenVertices = edge.HiddenVertices;
                    if (addOldVertexAsHidden) hiddenVertices.Add(vertex);
                    NetworkEdge newEdge = new NetworkEdge(newVertex, edge.Target, edge.IsGreedy, edge.Distance, hiddenVertices);
                    graph.AddEdge(newVertex, edge.Target, newEdge);
                }
                else
                {
                    hiddenVertices = new List<NetworkVertex>();
                    if (addOldVertexAsHidden) hiddenVertices.Add(vertex);
                    hiddenVertices.AddRange(edge.HiddenVertices);
                    NetworkEdge newEdge = new NetworkEdge(edge.Source, newVertex, edge.IsGreedy, edge.Distance, hiddenVertices);
                    graph.AddEdge(newEdge.Source, newVertex, newEdge);
                }
            }
            return newVertex;
        }

        /**
         * replaces one vertex by another for a network graph
         * copies all edges
         */
        public static bool ReplaceVertex(SimpleGraph<NetworkVertex, NetworkEdge> graph,
                NetworkVertex oldVertex, NetworkVertex newVertex)
        {
            // add new vertex
            graph.AddVertex(newVertex);
            // replace old edges
            //Set<NetworkEdge> oldEdges = graph.edgesOf(oldVertex);
            var oldEdges = graph.EdgesOf(oldVertex);
            foreach (NetworkEdge oldEdge in oldEdges)
            {
                NetworkEdge newEdge = NetworkEdge.ReplaceVertex(oldEdge, oldVertex, newVertex);
                if (newEdge.Source == newVertex)
                {
                    graph.AddEdge(newVertex, newEdge.Target, newEdge);
                }
                else
                {
                    graph.AddEdge(newEdge.Source, newVertex, newEdge);
                }
            }
            // remove old vertex
            return graph.RemoveVertex(oldVertex);
        }

        /**
         * Filters all vertices from a collection of vertices that lay in a specified collection of hexes
         */
        public static HashSet<NetworkVertex> GetVerticesByHexes(IEnumerable<NetworkVertex> vertices, ICollection<MapHex> hexes)
        {
            HashSet<NetworkVertex> hexVertices = new HashSet<NetworkVertex>();
            foreach (NetworkVertex vertex in vertices)
            {
                if (vertex.Hex != null && hexes.Contains(vertex.Hex))
                {
                    hexVertices.Add(vertex);
                }
            }
            return hexVertices;
        }

        /**
         * Returns all vertices for a specified hex
         */
        public static HashSet<NetworkVertex> GetVerticesByHex(IEnumerable<NetworkVertex> vertices, MapHex hex)
        {
            HashSet<NetworkVertex> hexVertices = new HashSet<NetworkVertex>();
            foreach (NetworkVertex vertex in vertices)
            {
                if (vertex.Hex != null && hex == vertex.Hex)
                {
                    hexVertices.Add(vertex);
                }
            }
            return hexVertices;
        }

        public static NetworkVertex GetVertexByIdentifier(IEnumerable<NetworkVertex> vertices, string identifier)
        {
            foreach (NetworkVertex vertex in vertices)
            {
                if (vertex.Identifier.Equals(identifier))
                {
                    return vertex;
                }
            }
            return null;
        }

        // #FIXME_GUI
        public static PointF? GetVertexPoint2D(HexMap map, NetworkVertex vertex)
        {
            if (vertex.IsVirtual) return default(PointF);//null;

            GUIHex guiHex = map.GetHex(vertex.Hex);
            if (vertex.IsMajor)
            {
                return guiHex.GetStopPoint2D(vertex.Stop);
            }
            else if (vertex.IsMinor)
            {
                return guiHex.GetStopPoint2D(vertex.Stop);
                //            return guiHex.getCenterPoint2D();
            }
            else if (vertex.IsSide)
            {
                // FIXME: Check if this still works
                return guiHex.GetSidePoint2D(vertex.Side);
            }
            else
            {
                return null;
            }
        }

        public static Rectangle? GetVertexMapCoverage(HexMap map, IEnumerable<NetworkVertex> vertices)
        {

            Rectangle? rectangle = null;

            // find coverage are of the vertices
            float minX = 0, minY = 0, maxX = 0, maxY = 0;
            foreach (NetworkVertex vertex in vertices)
            {
                PointF? point_nullable = GetVertexPoint2D(map, vertex);
                if (point_nullable != null)
                {
                    PointF point = (PointF)point_nullable;
                    if (minX == 0)
                    { // init
                      //rectangle = new Rectangle((int)point.X, (int)point.Y, 0, 0);
                        minX = point.X;
                        minY = point.Y;
                        maxX = minX; maxY = minY;
                    }
                    else
                    {
                        //rectangle.Add(point);
                        minX = Math.Min(minX, point.X);
                        minY = Math.Min(minY, point.Y);
                        maxX = Math.Max(maxX, point.X);
                        maxY = Math.Max(maxY, point.Y);
                    }
                }
            }
            rectangle = new Rectangle((int)minX, (int)minY, (int)(maxX - minX), (int)(maxY - minY));
            log.Info("Vertex Map Coverage minX=" + minX + ", minY=" + minY + ", maxX=" + maxX + ", maxY=" + maxY);
            //        Rectangle rectangle = new Rectangle((int)minX, (int)minY, (int)maxX, (int)maxY);
            log.Info("Created rectangle=" + rectangle);
            return (rectangle);
        }
    }
}

File: src/main/java/net/sf/rails/algorithms/NetworkVertex.java
package net.sf.rails.algorithms;

import java.awt.Rectangle;
import java.awt.geom.Point2D;
import java.util.*;

import net.sf.rails.game.*;
import net.sf.rails.ui.swing.hexmap.*;

import net.sf.rails.util.Util;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.jgrapht.graph.SimpleGraph;


public final class NetworkVertex implements Comparable<NetworkVertex> {

    private static final Logger log = LoggerFactory.getLogger(NetworkVertex.class);

    public enum VertexType {
        STATION,
        SIDE,
        HQ,
    }
    public enum StationType {
        MAJOR,
        MINOR,
        COALMINE,
        PASS
    }

    // vertex types and flag for virtual (thus not related to a rails object)
    private final VertexType type;
    private final boolean virtual;

    // vertex properties (for virtual vertexes)
    private final String virtualId;

    // general vertex properties
    private StationType stationType;
    private int value = 0;
    private boolean sink = false;
    private String stopName = null;
    private String mutexId = null;

    // references to rails objects, if not virtual
    private final MapHex hex;
    private final TrackPoint trackPoint;
    // only for station objects
    private final Stop stop;


    /** constructor for station on mapHex */
    public NetworkVertex(MapHex hex, Station station) {
        this.type = VertexType.STATION;
        this.hex = hex;
        this.trackPoint = station;
        this.stop = hex.getRelatedStop(station);
        if (stop != null) {
            log.debug("Found stop {}", stop);
        } else {
            log.debug("No stop found");
        }

        this.virtual = false;
        this.virtualId = null;
    }

    /** constructor for side on mapHex */
    public NetworkVertex(MapHex hex, HexSide side) {
        this.type = VertexType.SIDE;
        this.hex = hex;
        this.trackPoint = side;
        this.stop = null;

        this.virtual = false;
        this.virtualId = null;
    }

    /**  constructor for public company hq */
    public NetworkVertex(PublicCompany company) {
        this(VertexType.HQ, "HQ");
    }

    private NetworkVertex(VertexType type, String name) {
        this.type = type;
        this.hex = null;
        this.trackPoint = null;
        this.stop = null;

        this.virtual = true;
        this.virtualId = name;
    }

    /** factory method for virtual vertex
     */
    public static NetworkVertex getVirtualVertex(VertexType type, String name) {
        return new NetworkVertex(type, name);
    }

    void addToRevenueCalculator(RevenueCalculator rc, int vertexId) {
        rc.setVertex(vertexId, isMajor(), isMinor(), sink);
    }

    public String getIdentifier(){
        if (virtual) {
            return virtualId;
        } else {
            return hex.getId() + "." + trackPoint.getTrackPointNumber();
        }
    }

    public boolean isVirtual() {
        return virtual;
    }

    public boolean isStation(){
        return type == VertexType.STATION;
    }

    public boolean isSide(){
        return type == VertexType.SIDE;
    }

    public boolean isHQ(){
        return type == VertexType.HQ;
    }

    public VertexType getType() {
        return type;
    }

    public boolean isMajor(){
        return (stationType != null && stationType == StationType.MAJOR);
    }

    public boolean isMinor(){
        return (stationType != null && stationType == StationType.MINOR);
    }

    public boolean isMine( ) {
        return (stationType != null && stationType == StationType.COALMINE);
    }

    public boolean isPass() { return stationType != null && stationType == StationType.PASS; }

    public StationType getStationType() {
        return stationType;
    }

    public NetworkVertex setStationType(StationType stationType) {
        this.stationType = stationType;
        return this;
    }

    public int getValue() {
        return value;
    }

    public int getValueByTrain(NetworkTrain train) {
        int valueByTrain;
        if (isMajor()) {
            valueByTrain = value * train.getMultiplyMajors();
        } else if (isMinor()) {
            // FIXME: the next line is probably wrong or insufficient. See TrainType near line 85.
            if (train.ignoresMinors()) {
                valueByTrain = 0;
            } else {
                valueByTrain = value * train.getMultiplyMinors();
            }
        } else if (isMine()) { //Sofar Coal mines count nothing for the dividend income, the revenue from a mine is counted differently.
            valueByTrain = 0;
        } else {
            valueByTrain = value;
        }
        return valueByTrain;
    }

    public NetworkVertex setValue(int value) {
        this.value = value;
        return this;
    }

    public boolean isSink() {
        return sink;
    }

    public NetworkVertex setSink(boolean sink) {
        this.sink = sink;
        return this;
    }

    public String getStopName() {
        return stopName;
    }

    public String getMutexId() {
        return mutexId;
    }

    // getter for rails objects
    public MapHex getHex(){
        return hex;
    }

    public Station getStation(){
        if (type == VertexType.STATION) {
            return (Station) trackPoint;
        } else {
            return null;
        }
    }

    public HexSide getSide(){
        if (type == VertexType.SIDE) {
            return (HexSide) trackPoint;
        } else {
            return null;
        }
    }

    public Stop getStop() {
        return stop;
    }

    public boolean isOfType(VertexType vertexType, StationType stationType) {
        return (type == vertexType && (!isStation() || getStationType() == stationType));
    }

    /**
     * Initialize for rails vertexes
     * @param running true for train runs, false for tile or token lay allowances
     * @return true = can stay inside the network, false = has to be removed
     */
    public boolean initRailsVertex(PublicCompany company, boolean running) {
        // side vertices use the defaults, virtuals cannot use this function
        if (virtual || type == VertexType.SIDE) return true;

        // Only station remains
        Station station = (Station) trackPoint;

        log.debug("Init of vertex {}", this);
        // check if it has to be removed because it is run-to only
        // if company == null, then no vertex gets removed
        if (company != null && !stop.isRunToAllowedFor(company, running)
                && !stop.isRunThroughAllowedFor(company)) {
           log.info("Vertex is removed");
           return false;
        }

        // check if it is a major or minor
        if (stop.getScoreType() == Access.Score.MAJOR) {
            setStationType(StationType.MAJOR);
        } else if (stop.getScoreType() == Access.Score.MINOR) {
            setStationType(StationType.MINOR);
        } else if (stop.getScoreType() == Access.Score.NO) { // Used in 18EU Alpine variant
            setStationType(StationType.PASS); // Not sure if this is sensible for 18EU
        }

        // check if it is a sink
        if (company == null) { // if company == null, then all sinks are deactivated
            sink = false;
        } else {
            sink = !stop.isRunThroughAllowedFor(company);
        }

        // Changed 24/6/2020 by EV: separate the stop name (= city name) from the
        // mutexId. The former is only used in some modifiers, the latter governs
        // any loop restrictions.
        stopName = hex.getStopName();
        mutexId = stop.getMutexId();
        if (stop.getRelatedStation().getType() == Stop.Type.OFFMAP) {
            // For offmap hexes only, the city (or area) name doubles as a default mutexId.
            if (mutexId == null && StringUtils.isNotBlank(hex.getStopName()) ) {
                mutexId = hex.getStopName();
            }
        }


        // no removal
        return true;

    }

    public void setRailsVertexValue(Phase phase) {
        // side vertices and  virtuals cannot use this function
        if (virtual || type == VertexType.SIDE) return;

        // define value
        value = stop.getValueForPhase(phase);
    }


    @Override
    public String toString(){
        StringBuilder message = new StringBuilder();
        if (isVirtual())
            message.append(virtualId);
        else if (isStation())
            message.append(hex.getId()).append(".").append(((Station) trackPoint).getNumber());
        else if (isSide())
            message.append(hex.getId()).append(".").append(hex.getOrientationName((HexSide) trackPoint));
        else
            message.append("HQ");
        if (isSink())
            message.append("/*");
        return message.toString();
    }

    public int compareTo(NetworkVertex otherVertex) {
        return this.getIdentifier().compareTo(otherVertex.getIdentifier());
    }

    public static final class ValueOrder implements Comparator<NetworkVertex> {

        public int compare(NetworkVertex vA, NetworkVertex vB) {
            int result = -Integer.compare(vA.getValue(), vB.getValue()); // compare by value, descending
            if (result == 0)
                result = vA.compareTo(vB); // otherwise use natural ordering
            return result;
        }
    }

    /**
     *
     * @param graph network graph
     * @param company the company (with regard to values, sinks and removals)
     * @param phase the current phase (with regard to values)
     * @param running true for train runs, false for tile or token lay allowances
     */
    public static void initAllRailsVertices(NetworkGraph graph,
            PublicCompany company,  Phase phase, boolean running) {

        // store vertices for removal
        List<NetworkVertex> verticesToRemove = new ArrayList<>();
        for (NetworkVertex v:graph.getGraph().vertexSet()) {
            if (company != null) {
                if (!v.initRailsVertex(company, running)) {
                    verticesToRemove.add(v);
                    log.debug ("Vertex {} will be removed from the graph", v);
                }
            }
            if (phase != null) {
                v.setRailsVertexValue(phase);
            }
        }
        graph.getGraph().removeAllVertices(verticesToRemove);
    }

    /**
     * Returns the maximum positive value (lower bound zero)
     */
    public static int maxVertexValue(Collection<NetworkVertex> vertices) {
        int maximum = 0;
        for (NetworkVertex vertex:vertices) {
            maximum = Math.max(maximum, vertex.getValue());
        }
        return maximum;
    }


    /**
     * Return the sum of vertex values
     */
    public static int sum(Collection<NetworkVertex> vertices) {
        int sum = 0;
        for (NetworkVertex vertex:vertices) {
            sum += vertex.getValue();
        }
        return sum;
    }

    /**
     * Returns the number of specified vertex type in a vertex collection
     * If station then specify station type
     */
    public static int numberOfVertexType(Collection<NetworkVertex> vertices, VertexType vertexType, StationType stationType) {
        int number = 0;
        for (NetworkVertex vertex:vertices) {
            if (vertex.isOfType(vertexType, stationType)) number++;
        }
        return number;
    }

    /**
     * creates a new virtual vertex with identical properties and links
     */
    public static NetworkVertex duplicateVertex(SimpleGraph<NetworkVertex, NetworkEdge> graph,
            NetworkVertex vertex, String newIdentifier, boolean addOldVertexAsHidden) {
        // create new vertex
        NetworkVertex newVertex = NetworkVertex.getVirtualVertex(vertex.type, newIdentifier);
        // copy values
        newVertex.stationType = vertex.stationType;
        newVertex.value = vertex.value;
        newVertex.sink = vertex.sink;
        newVertex.stopName = vertex.stopName;
        newVertex.mutexId = vertex.mutexId;
        graph.addVertex(newVertex);
        // copy edges
        Set<NetworkEdge> edges = graph.edgesOf(vertex);
        for (NetworkEdge edge:edges) {
            List<NetworkVertex> hiddenVertices;
            if (edge.getSource() == vertex) {
                hiddenVertices = edge.getHiddenVertices();
                if (addOldVertexAsHidden) hiddenVertices.add(vertex);
                NetworkEdge newEdge = new NetworkEdge(newVertex, edge.getTarget(), edge.isGreedy(), edge.getDistance(), hiddenVertices);
                graph.addEdge(newVertex, edge.getTarget(), newEdge);
            } else {
                hiddenVertices = new ArrayList<>();
                if (addOldVertexAsHidden) hiddenVertices.add(vertex);
                hiddenVertices.addAll(edge.getHiddenVertices());
                NetworkEdge newEdge = new NetworkEdge(edge.getSource(), newVertex, edge.isGreedy(), edge.getDistance(), hiddenVertices);
                graph.addEdge(newEdge.getSource(), newVertex, newEdge);
            }
        }
        return newVertex;
    }

    /**
     * replaces one vertex by another for a network graph
     * copies all edges
     */
    public static boolean replaceVertex(SimpleGraph<NetworkVertex, NetworkEdge> graph,
            NetworkVertex oldVertex, NetworkVertex newVertex) {
        // add new vertex
        graph.addVertex(newVertex);
        // replace old edges
        Set<NetworkEdge> oldEdges = graph.edgesOf(oldVertex);
        for (NetworkEdge oldEdge:oldEdges) {
            NetworkEdge newEdge = NetworkEdge.replaceVertex(oldEdge, oldVertex, newVertex);
            if (newEdge.getSource() == newVertex) {
                graph.addEdge(newVertex, newEdge.getTarget(), newEdge);
            } else {
                graph.addEdge(newEdge.getSource(), newVertex, newEdge);
            }
        }
        // remove old vertex
        return graph.removeVertex(oldVertex);
    }

    /**
     * Filters all vertices from a collection of vertices that lay in a specified collection of hexes
     */
    public static Set<NetworkVertex> getVerticesByHexes(Collection<NetworkVertex> vertices, Collection<MapHex> hexes) {
        Set<NetworkVertex> hexVertices = new HashSet<>();
        for (NetworkVertex vertex:vertices) {
            if (vertex.getHex() != null && hexes.contains(vertex.getHex())) {
                hexVertices.add(vertex);
            }
        }
        return hexVertices;
    }

    /**
     * Returns all vertices for a specified hex
     */
    public static Set<NetworkVertex> getVerticesByHex(Collection<NetworkVertex> vertices, MapHex hex) {
        Set<NetworkVertex> hexVertices = new HashSet<>();
        for (NetworkVertex vertex:vertices) {
            if (vertex.getHex() != null && hex == vertex.getHex()) {
                hexVertices.add(vertex);
            }
        }
        return hexVertices;
    }

    public static NetworkVertex getVertexByIdentifier(Collection<NetworkVertex> vertices, String identifier) {
        for (NetworkVertex vertex:vertices) {
            if (vertex.getIdentifier().equals(identifier)) {
                return vertex;
            }
        }
        return null;
    }

    public static Point2D getVertexPoint2D(HexMap map, NetworkVertex vertex) {
        if (vertex.isVirtual()) return null;

        GUIHex guiHex = map.getHex(vertex.getHex());
        if (vertex.isMajor()) {
            return guiHex.getStopPoint2D(vertex.getStop());
        } else if (vertex.isMinor() || vertex.isPass()) {
            return guiHex.getStopPoint2D(vertex.getStop());
            //            return guiHex.getCenterPoint2D();
        } else if (vertex.isSide()) {
            // FIXME: Check if this still works
            return guiHex.getSidePoint2D(vertex.getSide());
        } else {
            return null;
        }
    }

    public static Rectangle getVertexMapCoverage(HexMap map, Collection<NetworkVertex> vertices) {

        Rectangle rectangle = null;

        // find coverage are of the vertices
        double minX=0,minY=0,maxX=0,maxY=0;
        for (NetworkVertex vertex:vertices) {
            Point2D point = getVertexPoint2D(map, vertex);
            if (point != null) {
                if (minX == 0) { // init
                    rectangle = new Rectangle((int)point.getX(), (int)point.getY(), 0, 0);
                    minX = point.getX();
                    minY = point.getY();
                    maxX = minX; maxY = minY;
                } else {
                    rectangle.add(point);
                    minX = Math.min(minX, point.getX());
                    minY = Math.min(minY, point.getY());
                    maxX = Math.max(maxX, point.getX());
                    maxY = Math.max(maxY, point.getY());
                }
            }
        }
        log.debug("Vertex Map Coverage minX={}, minY={}, maxX={}, maxY={}", minX, minY, maxX, maxY);
        //        Rectangle rectangle = new Rectangle((int)minX, (int)minY, (int)maxX, (int)maxY);
        log.debug("Created rectangle={}", rectangle);
        return (rectangle);
    }

    /* Added for 1826, as it turned out that the same vertex
     * could manifestate as different objects while iterating
     * over the vertices in the LinkedList behind DLLGraph.Segment.
     */
    public boolean equals (NetworkVertex otherVertex) {
        return toString().equals(otherVertex.toString());
    }
}
